control-plane/config-entries/controllersv2/meshconfig_controller.go

```diff
@@ -0,0 +1,300 @@
+
+	"github.com/hashicorp/consul-k8s/control-plane/api/common"
+	tenancy "github.com/hashicorp/consul-k8s/control-plane/connect-inject/common"
+	"github.com/hashicorp/consul-k8s/control-plane/connect-inject/constants"
+	"github.com/hashicorp/consul-k8s/control-plane/consul"
+	"github.com/hashicorp/consul-k8s/control-plane/namespaces"
+)
+
+const (
+	ConsulAgentError             = "ConsulAgentError"
+	ExternallyManagedConfigError = "ExternallyManagedConfigError"
+)
+
+// Controller is implemented by CRD-specific config-entries. It is used by
+// MeshConfigController to abstract CRD-specific config-entries.
+type Controller interface {
+	// Update updates the state of the whole object.
+	Update(context.Context, client.Object, ...client.UpdateOption) error
+	// UpdateStatus updates the state of just the object's status.
+	UpdateStatus(context.Context, client.Object, ...client.SubResourceUpdateOption) error
+	// Get retrieves an obj for the given object key from the Kubernetes Cluster.
+	// obj must be a struct pointer so that obj can be updated with the response
+	// returned by the Server.
+	Get(ctx context.Context, key client.ObjectKey, obj client.Object, opts ...client.GetOption) error
+	// Logger returns a logger with values added for the specific controller
+	// and request name.
+	Logger(types.NamespacedName) logr.Logger
+}
+
+// MeshConfigController is a generic controller that is used to reconcile
+// all resource types, e.g. TrafficPermissions, ProxyConfiguration, etc., since
+// they share the same reconcile behaviour.
+type MeshConfigController struct {
+	// ConsulClientConfig is the config for the Consul API client.
+	ConsulClientConfig *consul.Config
+
+	// ConsulServerConnMgr is the watcher for the Consul server addresses.
+	ConsulServerConnMgr consul.ServerConnectionManager
+
+	common.ConsulTenancyConfig
+}
+
+// ReconcileEntry reconciles an update to a resource. CRD-specific controller's
+// call this function because it handles reconciliation of config entries
+// generically.
+// CRD-specific controller should pass themselves in as updater since we
+// need to call back into their own update methods to ensure they update their
+// internal state.
+func (r *MeshConfigController) ReconcileEntry(ctx context.Context, crdCtrl Controller, req ctrl.Request, meshConfig common.MeshConfig) (ctrl.Result, error) {
+	logger := crdCtrl.Logger(req.NamespacedName)
+	err := crdCtrl.Get(ctx, req.NamespacedName, meshConfig)
+	if k8serr.IsNotFound(err) {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	} else if err != nil {
+		logger.Error(err, "retrieving resource")
+		return ctrl.Result{}, err
+	}
+
+	// Create Consul resource service client for this reconcile.
+	resourceClient, err := consul.NewResourceServiceClient(r.ConsulServerConnMgr)
+	if err != nil {
+		logger.Error(err, "failed to create Consul resource client", "name", req.Name, "ns", req.Namespace)
+		return ctrl.Result{}, err
+	}
+
+	// TODO: add finalizers
```

Comment:
TBH, I ran out of time here without having a chance to investigate. If someone thinks this is essential functionality for the RC, let me know.

Reply:
nm, this was important for deleting the object